src/util/timer.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::time::Duration;
+use util::time::Instant;
+use std::collections::BinaryHeap;
+
+pub struct TimeoutTask<T> {
+    next_tick: Instant,
+    timeout: Duration,
+    task: T,
+}
+
+impl<T> TimeoutTask<T> {
+    pub fn restore(self, timer: &mut Timer<T>) {
+        timer.pending.push(self);
+    }
+}
+
+impl<T> AsRef<T> for TimeoutTask<T> {
+    fn as_ref(&self) -> &T {
+        &self.task
+    }
+}
+
+impl<T> AsMut<T> for TimeoutTask<T> {
+    fn as_mut(&mut self) -> &mut T {
+        &mut self.task
+    }
+}
+
+impl<T> PartialEq for TimeoutTask<T> {
+    fn eq(&self, other: &TimeoutTask<T>) -> bool {
+        if self.next_tick == other.next_tick {
+            return true;
+        }
+        false
+    }
+}
+
+impl<T> Eq for TimeoutTask<T> {}
+
+impl<T> PartialOrd for TimeoutTask<T> {
+    fn partial_cmp(&self, other: &TimeoutTask<T>) -> Option<Ordering> {
+        Some(if self.next_tick > other.next_tick {
+            Ordering::Less
+        } else if self.next_tick < other.next_tick {
+            Ordering::Greater
+        } else {
+            Ordering::Equal
+        })
+    }
+}
+
+impl<T> Ord for TimeoutTask<T> {
+    fn cmp(&self, other: &TimeoutTask<T>) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+pub struct Timer<T> {
+    now: Instant,
+    tick: Duration,
+    pending: BinaryHeap<TimeoutTask<T>>,
+}
+
+impl<T> Timer<T> {
+    fn align_up_instant(&self, ins: Instant) -> Instant {
+        let mut next = self.now + self.tick;
+        let now = Instant::now_coarse();
+        loop {
+            if next < ins || next < now {
+                next += self.tick;
+                continue;
+            }
+            break next;
+        }
+    }
+
+    fn deliver(&self, mut task: TimeoutTask<T>, buffer: &mut Vec<TimeoutTask<T>>) {
+        task.next_tick = self.align_up_instant(task.next_tick + task.timeout);
+        buffer.push(task);
+    }
+
+    /// Create a new timer with a specified `tick` which means
+    /// the min timeout when the `Timer` calls `next_timeout`.
+    pub fn new(tick: Duration) -> Self {
```

Comment:
Why `tick` is necessary?

Reply:
We can use it to avoid get too small timeout. For example, if we have a task should be ticked for every 7 ms, and another should be ticked for 8 ms. So the `Timer` sometimes will return a timeout `1 ms`. `tick` means the min timeout the `Timer` can return.

Reply:
At the cost of incorrect timeout? Does TiKV have such use case?

Reply:
`tick` can also deal with this case:

If the waste too much time to do other thing, the first `timeout` in pending queue maybe already expired. We can align it up to `tick` to avoid `zero timeout`.